Start implementing stream internal slots and algos

gterzian opened the pull request:
<!-- Please describe your changes on the following line: -->


---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

gterzian committed:
implement basic internal slots, with todos

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
enum for controller

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
re-implement native controller methods

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

wusyong commented on components/script/dom/readablestream.rs:
```diff
@@ -90,266 +118,196 @@ impl ReadableStream {
-            Box::new(ReadableStream::new_inherited(external_underlying_source)),
-            global,
-        )
+    fn new(global: &GlobalScope, controller: Controller) -> DomRoot<ReadableStream> {
```
Step 3 in https://streams.spec.whatwg.org/#rs-constructor will need to create a `ReadableStream` but controller isn't created yet at that time. The controller is set in https://streams.spec.whatwg.org/#set-up-readable-stream-default-controller.

gterzian replied:
Right, that step is still called from within the constructor, but if necessary we can wrap the controller in a `MutNullableDom`.

wusyong commented on components/script/dom/readablestreamdefaultcontroller.rs:
```diff
@@ -2,18 +2,124 @@
+    }
+
+    /// <https://streams.spec.whatwg.org/#ref-for-abstract-opdef-readablestreamcontroller-pullsteps>
+    pub fn perform_pull_steps(&self, read_request: ReadRequest) {
```
My issue is more about [`pullAlgorithm`](https://streams.spec.whatwg.org/#readablestreamdefaultcontroller-pullalgorithm) and similar internal slots. Because if you search `Let pullAlgorithm`, you can see there are multiple ways to create it. Do you have any suggestions on how to implement it?

jdm replied:
Using a trait object or function pointer would be my first attempt.

gterzian replied:
I've added a commit that calls into the pull algorithm.

wusyong replied:
Ah I guess that means we will need to implement the algos to other vairants.

gterzian committed:
add calling into pull algo

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
more details on chunk enqueuing

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

wusyong reviewed (approved):
LGTM now, but I still don't have the permissions. Could I get some access to review and add labels?

gterzian committed:
add fulfill read request, clean-up warnings

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
read request and reader typing

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
allow for more than one non-native underlying source type

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add todo for should pull

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

gterzian committed:
add underlying source dom struct container

Signed-off-by: gterzian <2792687+gterzian@users.noreply.github.com>

servo-wpt-sync commented:
🛠 These changes could not be applied onto the latest upstream WPT. Servo's copy of the Web Platform Tests may be out of sync.

gterzian committed:
remove rc around source type

gterzian committed:
add default controller init in stream constructor

gterzian commented on components/script/dom/underlyingsourcecontainer.rs:
```diff
@@ -0,0 +1,138 @@
+    ) -> DomRoot<UnderlyingSourceContainer> {
+        // Setting the prototype of the underlying source dict on the
+        // `UnderlyingSourceContainer` for later use in Call_.
+        // TODO: is this a good idea?
```
cc @jdm

gterzian replied:
I ended-up removing this entirely for now, it was crashing.

wusyong replied:
I still think we probably should expand the codegen to support underlying source that can call `Call`. I guess this is because it's the first dictionary that has callbacks.

jdm replied:
One thing that occurs to me—I believe the DomObject bound us only needed to call wrap_call_this_object, which only uses it to call the reflector method so it can get a JS object pointer. If we add a new trait to return the JS object directly instead, we should not need to play around with dictionary prototypes, just generate an impl for the new trait instead.

gterzian replied:
thank you, have included your example.

gterzian committed:
setup source container with prototype of source dict

gterzian committed:
clean-up docs, dispatch of controller in pull algo call

gterzian committed:
turn off SM streams

gterzian committed:
remove prototype setting on underlying source container

gterzian committed:
fix read request promise resolving

gterzian commented:
Ok, I also turned-off the old streams, and as a kind of milestone, https://github.com/servo/servo/blob/496ce717c5343984e0f3da00d223604eff03893c/tests/wpt/tests/fetch/api/basic/stream-response.any.js is now passing again(which covers part of the "native" streaming I handled in this PR). 

So I'd say let review and merge this one, so that work can resume in parallel on:
- The BYOB controller/reader part(note we were [not running tests](https://github.com/servo/servo/blob/496ce717c5343984e0f3da00d223604eff03893c/tests/wpt/include.ini#L217) for that part in the previous implementation, even though SM seems the implement it)
- Various TODOs in the default controller/reader. 
- other stuff?

gterzian committed:
tidy

gterzian committed:
clean-up js conversions in read req handlers

jdm commented:
I'll try to take a look at this tonight!

sagudev commented:
> note we were [not running tests](https://github.com/servo/servo/blob/496ce717c5343984e0f3da00d223604eff03893c/tests/wpt/include.ini#L217) for that part in the previous implementation, even though SM seems the implement it

Doesn't that mean that readable-streams test's are run (not skipped)?

jdm commented on components/script/dom/readablestream.rs:
```diff
@@ -182,7 +203,15 @@ impl ReadableStream {
+        match self.reader {
+            ReaderType::Default(ref reader) => {
+                let Some(reader) = reader.get() else {
+                    panic!("Attempt to read stream chunk without having acquired a reader.");
```
nit: s/read stream chunk/add read request/

jdm commented on components/script/dom/readablestream.rs:
```diff
@@ -235,39 +264,54 @@ impl ReadableStream {
+        match self.reader {
+            ReaderType::Default(ref reader) => {
+                let Some(rooted_reader) = reader.get() else {
+                    panic!("Attempt to read stream chunk without having acquired a reader.");
```
nit: s/read stream chunk/stop reading/

jdm commented on components/script/dom/underlyingsourcecontainer.rs:
```diff
@@ -75,6 +78,7 @@ impl UnderlyingSourceContainer {
         let cx = GlobalScope::get_cx();
         rooted!(in(*cx) let mut underlying_source_dict = UndefinedValue());
         if let UnderlyingSourceType::Js(ref js_source) = underlying_source_type {
+            #[allow(unsafe_code)]
```
nit: it makes more sense to have this annotation at the method level.

jdm commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -29,18 +34,47 @@ pub enum ReadRequest {
 impl ReadRequest {
     /// <https://streams.spec.whatwg.org/#read-request-chunk-steps>
+    #[allow(unsafe_code)]
+    #[allow(crown::unrooted_must_root)]
```
Why is this annotation required?

gterzian replied:
I was able to remove it by avoiding to assign the read result to a temporary variable.

gterzian replied:
I was able to remove it by avoiding to assign the read result to a temporary variable.

jdm commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -29,18 +34,47 @@ pub enum ReadRequest {
+                rooted!(in(*cx) let mut rval = UndefinedValue());
+                let result = RootedTraceableBox::new(Heap::default());
+                result.set(*rval);
+                let result = ReadableStreamReadResult {
```
This value does not appear to be used.

gterzian commented:
> Doesn't that mean that readable-streams test's are run (not skipped)?

yes but the byte stream stuff is in a separate folder, at https://github.com/servo/servo/tree/main/tests/wpt/tests/streams/readable-byte-streams, so we were never running them.

gterzian committed:
add queue with sizes concept

gterzian commented on components/script/dom/readablestreamdefaultcontroller.rs:
```diff
@@ -2,18 +2,92 @@
+
+    pub fn enqueue_chunk(&self, mut chunk: Vec<u8>) {
+        let mut buffer = self.buffer.borrow_mut();
+        chunk.append(&mut buffer);
```
Yes that seems to be the case. 

On this and the comment below: I looked into this and actually we should re-structure this queue concept to allow to any JS value being a chunk, a first step of which I have done in the last commit. There is still a "native" chunk variant, which I guess we could optimize.

gterzian committed:
use dom in pull promise handlers

gterzian commented on components/script/dom/readablestreamdefaultcontroller.rs:
```diff
@@ -6,18 +6,55 @@ use std::cell::{Cell, RefCell};
+/// <https://streams.spec.whatwg.org/#readable-stream-default-controller-call-pull-if-needed>
+#[derive(Clone, JSTraceable, MallocSizeOf)]
+struct PullAlgorithmRejectionHandler {
+    controller: DomRoot<ReadableStreamDefaultController>,
```
How about using `Dom`? See last commit. Since `Trusted` is atomic and we don't need this here...

gterzian commented on components/script/dom/readablestreamdefaultcontroller.rs:
```diff
@@ -88,6 +99,27 @@ impl ReadableStreamDefaultController {
+    }
+
+    /// Return bytes synchronously if the stream has all data in memory.
+    pub fn get_in_memory_bytes(&self) -> Option<Vec<u8>> {
```
Yes I think we can look into something like that, this method is only used in `body.rs`.

jdm commented on components/script/dom/readablestreamdefaultcontroller.rs:
```diff
@@ -28,8 +29,9 @@ use crate::script_runtime::{JSContext, JSContext as SafeJSContext};
+#[allow(crown::unrooted_must_root)]
 struct PullAlgorithmFulfillmentHandler {
-    controller: DomRoot<ReadableStreamDefaultController>,
+    controller: Dom<ReadableStreamDefaultController>,
```
I don't think this is safe, since these structs aren't visible to the GC. That's why I recommend Trusted here.

gterzian replied:
> That's why I recommend Trusted here.
I just tried it and it breaks `Callback: JSTraceable + MallocSizeOf `. 

>  these structs aren't visible to the GC

But they end-up on the `PromiseNativeHandler`, which is a dom_struct.

gterzian replied:
I've strengthened this by keeping the controller rooted while the native promise handler is setup.

jdm committed:
Demonstrate using dictionary as callback this object.

jdm commented:
https://github.com/servo/servo/commit/b65ba8df924db6d0f2263798c6d8810dd953b64f is how I would support dictionary callback values.

gterzian committed:
move value with size to a struct

gterzian committed:
fmt

gterzian committed:
put readable stream state in a cell

gterzian committed:
nits in expectations

gterzian committed:
remove allow unroot by passing read result directly to promise resolving

gterzian commented on components/script/dom/readablestreamdefaultreader.rs:
```diff
@@ -2,25 +2,50 @@
+    /// <https://streams.spec.whatwg.org/#readablestreamgenericreader-stream>
+    stream: Dom<ReadableStream>,
+
+    #[ignore_malloc_size_of = "Rc is hard"]
```
The read request contains a `Rc<Promise>`.

gterzian committed:
tidy

gterzian committed:
root default controller inside call_pull_if_needed

wusyong commented:
> @wusyong @Taym95 I will address the remaining points from the review, and then I think we need to merge and work in parallel, because I am starting to run into some complicated "how to do this in JS" stuff that warrant PRs in and of themselves...

I thought this PR would be merged earlier so I was waiting for it. Byte controller and reader is pretty similar to default one. So I believe once we have done the default ones, the byte part will be faster. I'll review the PR again pretty soon.

wusyong reviewed (approved):
@jdm @gterzian Could we merge this today? I believe @gterzian [mentioned ](https://servo.zulipchat.com/#narrow/stream/263398-general/topic/Re-implement.20ReadableStream/near/451703639)we could do more aggressive mergine since this is targeted to the feature branch.
There are some trivial improvements like [EnqueuedValue](https://github.com/servo/servo/pull/32730/files#diff-09888dfa911f121670b8b905131455cf548d1816bf65bdeff76fb00f9e55e743R70) and maybe QueueWithSize should share with byob controller. But I think it could wait for there's PR for such controller implementation.

Edit: I also hope we can update the feature branch to catch up main branch since there are some more commit over weeks.

gterzian commented:
Ok let's do it! @jdm I have addressed all of your comments, although I still had a few questions(about the rooting), but this had been noted in a TODO as needed to revisit...
